File: backend/src/main/java/com/flowlink/execution/config/ExecutionProperties.java
package com.flowlink.execution.config;

import org.springframework.boot.context.properties.ConfigurationProperties;

import java.util.List;

/** flowlink.execution.* 설정 바인딩. 값이 없으면 안전한 기본값으로 채운다. */
@ConfigurationProperties(prefix = "flowlink.execution")
public record ExecutionProperties(
        Http http,
        Ssrf ssrf,
        Capture capture,
        int maxNodesPerRun
) {
    public ExecutionProperties {
        if (http == null) {
            http = new Http(5000, 30000, 5_242_880L);
        }
        if (ssrf == null) {
            ssrf = new Ssrf(true, true, false, List.of("169.254.169.254"), List.of("http", "https"));
        }
        if (capture == null) {
            capture = new Capture(false);
        }
        if (maxNodesPerRun <= 0) {
            maxNodesPerRun = 200;
        }
    }

    /**
     * 실행 로그 캡처 정책(redaction). 기본은 deny-by-default — 요청/응답 본문은 저장하지 않는다.
     * (본문엔 Authorization 헤더·토큰 등 시크릿이 섞일 수 있어 기본 미저장)
     */
    public record Capture(boolean requestResponseBodies) {
    }

    public record Http(int connectTimeoutMs, int readTimeoutMs, long maxResponseBytes) {
        public Http {
            if (connectTimeoutMs <= 0) {
                connectTimeoutMs = 5000;
            }
            if (readTimeoutMs <= 0) {
                readTimeoutMs = 30000;
            }
            if (maxResponseBytes <= 0) {
                maxResponseBytes = 5_242_880L;
            }
        }
    }

    public record Ssrf(
            boolean enabled,
            boolean blockPrivateNetworks,
            boolean allowLoopback,        // 로컬 배포용: true 면 localhost/127.0.0.1/::1 허용(사설망은 여전히 차단)
            List<String> blockedHosts,
            List<String> allowedSchemes
    ) {
        public Ssrf {
            if (blockedHosts == null) {
                blockedHosts = List.of();
            }
            if (allowedSchemes == null || allowedSchemes.isEmpty()) {
                allowedSchemes = List.of("http", "https");
            }
        }
    }
}

File: backend/src/main/kotlin/com/flowlink/execution/config/ExecutionProperties.kt
package com.flowlink.execution.config

import org.springframework.boot.context.properties.ConfigurationProperties

/** flowlink.execution.* 설정 바인딩. 값이 없으면 안전한 기본값으로 채운다. */
@ConfigurationProperties(prefix = "flowlink.execution")
class ExecutionProperties(
    http: Http?,
    maxNodesPerRun: Int = 0,
    worker: Worker? = null,
) {
    val http: Http = http ?: Http(5000, 30000, 5_242_880L)
    val maxNodesPerRun: Int = if (maxNodesPerRun <= 0) 200 else maxNodesPerRun
    val worker: Worker = worker ?: Worker()

    /** 실행 워커 풀(비동기 실행/재개 연속 실행 전용). 큐 초과 제출은 429 로 거절. */
    class Worker(poolSize: Int = 0, queueCapacity: Int = 0) {
        val poolSize: Int = if (poolSize <= 0) 8 else poolSize
        val queueCapacity: Int = if (queueCapacity <= 0) 100 else queueCapacity
    }

    class Http(
        connectTimeoutMs: Int = 0,
        readTimeoutMs: Int = 0,
        maxResponseBytes: Long = 0,
    ) {
        val connectTimeoutMs: Int = if (connectTimeoutMs <= 0) 5000 else connectTimeoutMs
        val readTimeoutMs: Int = if (readTimeoutMs <= 0) 30000 else readTimeoutMs
        val maxResponseBytes: Long = if (maxResponseBytes <= 0) 5_242_880L else maxResponseBytes
    }
}
